### PR TITLE
Storing comments

### PR DESCRIPTION
**Pull request by kofimeighan**
I added the user-submitted locations dropdown for the statistics page and populates the dropdown using the served JSON data from the server.
Screenshot 2020-07-06 at 12 51 56 PM

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/MapServlet.java [hunk 1]
```diff
@@ -40,32 +40,52 @@
...
-    pinEntity.setProperty("TimeStamp", timeStamp);
+    pinEntity.setProperty(NAME, name);
+    pinEntity.setProperty(PHONE, phone);
+    pinEntity.setProperty(LOCATION, location);
+    pinEntity.setProperty(DESCRIPTION, description);
+    pinEntity.setProperty(TIME_STAMP, timeStamp);
     datastore.put(pinEntity);
 
     response.setContentType("text/html");
-    response.getWriter().println("TEST");
```
Comment (kofimeighan) [inline portfolio/src/main/java/com/google/sps/servlets/MapServlet.java]: @wangtong-g Here I have deleted all the debugging points within the production code

File: portfolio/src/main/java/com/google/sps/servlets/MapServlet.java [hunk 2]
```diff
@@ -40,32 +40,52 @@
...
+  private static final String ID = "id";
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    Query query = new Query(TABLE_NAME);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    List<UserComment> userCommentsList = new ArrayList<UserComment>();
```
Comment (wangtong-g) [inline portfolio/src/main/java/com/google/sps/servlets/MapServlet.java]: Nit: using "userComments" as variable name should be good enough.

File: portfolio/src/main/java/com/google/sps/servlets/MapServlet.java [hunk 3]
```diff
@@ -40,32 +40,52 @@
...
+      String phone = (String) entity.getProperty(PHONE);
+      String location = (String) entity.getProperty(LOCATION);
+      String description = (String) entity.getProperty(DESCRIPTION);
+      long timeStamp = (long) entity.getProperty(TIME_STAMP);
+      long id = entity.getKey().getId();
+
+      UserComment userComment = new UserComment(name, phone, location, description, timeStamp, id);
+      userCommentsList.add(userComment);
+    }
+    System.out.println(userCommentsList);
```
Comment (wangtong-g) [inline portfolio/src/main/java/com/google/sps/servlets/MapServlet.java]: Same here, prefer logging instead of print.

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+package com.google.sps.servlets;
+
+public final class UserComment {
+  private final String name;
+  private final String phone;
+  private final String location;
+  private final String description;
+  private final long timeStamp;
+  private final long id;
```
Comment (wangtong-g) [inline portfolio/src/main/java/com/google/sps/servlets/UserComment.java:23]: What is the id here?

File: portfolio/src/main/webapp/index.html [hunk 5]
```diff
@@ -377,13 +378,14 @@ <h2 class="text-uppercase">George Floyd</h2>
                     while Lane, Keung, and Thao face charges of aiding and abetting second-degree murder and
                     manslaughter.
                   </p>
-                  <p><b>Justice for George Floyd</b></p>
+                  <hr />
```
Comment (wangtong-g) [inline portfolio/src/main/webapp/index.html:389]: Should be `<hr>.`

File: portfolio/src/main/webapp/index.html [hunk 6]
```diff
@@ -377,13 +378,14 @@ <h2 class="text-uppercase">George Floyd</h2>
...
+                  <h4><b>How to Help</b></h4>
                   <p>
                     <b>Donate Here: </b
                     ><a class="interactive-pics" href="https://www.gofundme.com/f/georgefloyd"
-                      >Offical George Floyd Memorial Fund</a
+                      >Official George Floyd Memorial Fund</a
                     ><br />
-                    <b>Sign Petition Here: </b
+                    <b>Sign a Petition Here: </b
                     ><a
```
Comment (wangtong-g) [inline portfolio/src/main/webapp/index.html]: This section needs some formatting.

File: portfolio/src/main/webapp/index.html [hunk 7]
```diff
@@ -412,25 +414,51 @@ <h2 class="text-uppercase">George Floyd</h2>
...
+                    grabbed a .357 Magnum handgun and shotgun, and chased Mr. Arbery in a pickup truck. The McMichael's
+                    attempted to cut Mr. Arbery off, but were unsuccessful as he began to <b>run for his life</b>. As he
+                    ran, he passed the truck of William Bryan Jr., the McMichael's neighbor, who recorded and eventually
+                    joined the chase. Mr. Arbery ran until he was trapped between the two trucks, and Travis McMichael
+                    got out of his truck with the shotgun. Mr. Arbery fought for his life when he was
+                    <b>shot 3 times</b> by Travis, followed by the yelling of a racial slur. Mr. Arbery stumbled and
+                    fell in the middle of the street, taking his last breath. All 3 men have since been arrested 2
+                    months later and indicted on 9 counts, including felony murder.
+                  </p>
+                  <hr />
```
Comment (wangtong-g) [inline portfolio/src/main/webapp/index.html:459]: `<hr>`

File: portfolio/src/main/webapp/index.html [hunk 8]
```diff
@@ -412,25 +414,51 @@ <h2 class="text-uppercase">George Floyd</h2>
...
+                  <p>
+                    <b>Donate Here: </b
+                    ><a class="interactive-pics" href="https://www.gofundme.com/f/i-run-with-maud">I Run With Maud</a
+                    ><br />
+                    <b>Sign Petitions Here: </b
+                    ><a
+                      class="interactive-pics"
+                      href="https://www.change.org/p/liberty-county-distric-attorney-justice-for-ahmaud-arbery"
+                      >Justice for Ahmaud Arbery</a
+                    >,
```
Comment (wangtong-g) [inline portfolio/src/main/webapp/index.html]: It's weird to put ">" next line.

File: portfolio/src/main/webapp/index.html [hunk 9]
```diff
@@ -412,25 +414,51 @@ <h2 class="text-uppercase">George Floyd</h2>
...
+                    <a
+                      class="interactive-pics"
+                      href="https://www.change.org/p/district-attorney-tom-durden-justice-for-ahmaud-arbery-i-run-with-maud"
+                      >Justice for Ahmaud Arbery! I Run with Maud!</a
+                    >
```
Comment (wangtong-g) [inline portfolio/src/main/webapp/index.html]: This formatting is weird too.

File: portfolio/src/main/webapp/index.html [hunk 10]
```diff
@@ -450,25 +478,44 @@ <h2 class="text-uppercase">Ahmaud Arbery</h2>
...
+                    Upon arrival, the officers began to use a battering ram to enter Ms. Taylor's apartment, while she
+                    and her boyfriend were in bed. The loud banging led to Kenneth Walker, Ms. Taylor's boyfriend,
+                    grabbed his own weapon and fired the gun at the officers. The officers retaliated and shot Ms.
+                    Taylor
+                    <b>at least 8 times</b>. The officers were believed to have been investigating 2 men that were
+                    selling drugs far from Ms. Taylor's home and no drugs were found in her apartment. Hankison was
+                    fired from the Louisville Metro Police Department while Mattingly and Cosgrove have been placed on
+                    administrative reassignment. Currently, <b>none of the officers face any criminal charges</b>.
+                  </p>
+                  <hr />
```
Comment (wangtong-g) [inline portfolio/src/main/webapp/index.html:534]: `<hr>`

File: portfolio/src/main/webapp/index.html [hunk 11]
```diff
@@ -450,25 +478,44 @@ <h2 class="text-uppercase">Ahmaud Arbery</h2>
...
+                  <h4>How to Help</h4>
+                  <p>
+                    <b>Donate Here: </b
+                    ><a class="interactive-pics" href="https://www.gofundme.com/f/9v4q2-justice-for-breonna-taylor"
+                      >Justice for Breonna Taylor (official)(#BREEWAYY)</a
+                    ><br />
+                    <b>Sign a Petition Here: </b
+                    ><a class="interactive-pics" href="https://www.change.org/p/andy-beshear-justice-for-breonna-taylor"
+                      >Justice for Breonna Taylor</a
+                    >
```
Comment (wangtong-g) [inline portfolio/src/main/webapp/index.html]: Please fix formatting.

File: portfolio/src/main/webapp/js/custom-script.js [hunk 12]
```diff
@@ -135,3 +126,16 @@ function insertSearch() {
 
   document.getElementById('mainNav').appendChild(searchBar);
 }
+
+async function addUserComments() {
+  const response = await fetch('/map');
```
Comment (wangtong-g) [inline portfolio/src/main/webapp/js/custom-script.js]: Why the endpoint of "user comments" is named as "map"?

**Review by wangtong-g — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/js/custom-script.js (hunk 12 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/js/custom-script.js]: I store the user-submitted data in the map endpoint and serve the results through the same servlet. I only named it 'map' to denote what the servlet is being used for.

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java (hunk 4 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserComment.java:23]: The Key that the specific entity is saved under within the database.

File: portfolio/src/main/webapp/index.html (hunk 11 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/index.html]: These changes were merged from master so I didn't review them within this pull request. In addition, when I run make pretty it reformats the tags to be on the next line. Should we side with make pretty or should I manually fix them?

File: portfolio/src/main/webapp/js/custom-script.js (hunk 12 above)
Comment (wangtong-g) [reply portfolio/src/main/webapp/js/custom-script.js]: So what does map endpoint do?

File: portfolio/src/main/webapp/index.html (hunk 11 above)
Comment (wangtong-g) [reply portfolio/src/main/webapp/index.html]: Unless style guide permits (I can't find this style in it), we have to manually fix this. We can't rely on the formatting tool without manual review.

File: portfolio/src/main/webapp/js/custom-script.js (hunk 12 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/js/custom-script.js]: Stores the user-submitted info into the database and serves it back to the javascript when requested

File: portfolio/src/main/webapp/index.html (hunk 8 above)
Comment (briafassler9) [reply portfolio/src/main/webapp/index.html]: This was more on my end. Whenever I run make pretty, it automatically formats the code and the ">" being on the next line is a result of that command.

File: portfolio/src/main/webapp/index.html (hunk 11 above)
Comment (briafassler9) [reply portfolio/src/main/webapp/index.html]: When I manually fix the formatting, using the make pretty command changes it back to the odd formatting. Should we only use make validate in the future? Any use of make pretty will result in the odd formatting.

File: portfolio/src/main/webapp/index.html [hunk 13]
```diff
@@ -377,18 +378,18 @@ <h2 class="text-uppercase">George Floyd</h2>
...
-                    ><br />
-                    <b>Sign Petition Here: </b
-                    ><a
+                    <b>Donate Here: </b>
+                    <a class="interactive-pics" href="https://www.gofundme.com/f/georgefloyd">
+                    Official George Floyd Memorial Fund</a>
+                    <br/>
+                    <b>Sign a Petition Here: </b>
+                    <a
                       class="interactive-pics"
```
Comment (kofimeighan) [inline portfolio/src/main/webapp/index.html]: Here is where the lint check will throw errors.

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java [hunk 14]
```diff
@@ -32,40 +32,59 @@
...
+    PreparedQuery results = datastore.prepare(query);
+
+    List<UserComment> userComments = new ArrayList<UserComment>();
+    for (Entity entity : results.asIterable()) {
+      String name = (String) entity.getProperty(NAME);
+      String phone = (String) entity.getProperty(PHONE);
+      String location = (String) entity.getProperty(LOCATION);
+      String description = (String) entity.getProperty(DESCRIPTION);
+      long timeStamp = (long) entity.getProperty(TIME_STAMP);
+      long id = entity.getKey().getId();
```
Comment (wangtong-g) [inline portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: Do you need to pass the id back to user?

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java [hunk 15]
```diff
@@ -32,40 +32,59 @@
...
+    pinEntity.setProperty(NAME, name);
+    pinEntity.setProperty(PHONE, phone);
+    pinEntity.setProperty(LOCATION, location);
+    pinEntity.setProperty(DESCRIPTION, description);
+    pinEntity.setProperty(TIME_STAMP, timeStamp);
     datastore.put(pinEntity);
 
     response.setContentType("text/html");
-    response.getWriter().println("TEST");
     response.sendRedirect("/statistics.html");
```
Comment (wangtong-g) [inline portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: After user post a location on map, why redirecting to stats page?

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java (hunk 14 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: Not necessarily, but I save the ID to the UserComment object just in case I will need to retrieve a specific comment in the future.

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java (hunk 15 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: Once the form is submitted, it would redirect the user to the /submitted-comments endpoint that is empty. In order to prevent this redirection, I redirect the user back to /statistics.html.

Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: re: the action="/submitted-comments" value in the form tag

Comment (wangtong-g) [reply portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: Then we should figure out why posting a new location will redirect user instead of having this workaround.

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java [hunk 16]
```diff
@@ -32,40 +32,59 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-@WebServlet("/map")
-public class MapServlet extends HttpServlet {
+@WebServlet("/submitted-comments")
```
Comment (wangtong-g) [inline portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: I tried very hard to accept this name - "submitted-comments". But every time I read it, it confuses me. Because it's not a comment. It's a location user submitted on map.

File: portfolio/src/main/webapp/index.html [hunk 17]
```diff
@@ -344,7 +344,7 @@ <h3 class="section-subheading text-muted">Click on each picture to learn more.</
     </footer>
     <!-- Martyr Modals-->
     <!-- TODO(briafassler/kofimeighan): finish inserting information for interactive pictures-->
-    <!-- Modal 1-->
+    <!-- Modal 1: George Floyd-->
```
Comment (kennzoid) [inline portfolio/src/main/webapp/index.html]: Is this PR also adding stuff to the home page as well? Or was this just formatting stuff? Or is this diffing against the wrong base branch?

Since this change seems unrelated to 'Storing comments', it should probably be in its own PR.

File: portfolio/src/main/webapp/js/custom-script.js [hunk 18]
```diff
@@ -61,6 +60,7 @@ function onLoad() {
   loadMap();
   populateDropdown(martyrData, 'martyr-dropdown-menu');
   populateDropdown(ipData, 'IP-dropdown-menu');
+  addUserComments();
```
Comment (kennzoid) [inline portfolio/src/main/webapp/js/custom-script.js]: It would be clearer if you also populated the user submitted dropdown here instead of within the helper.

For example:
```javascript
userData = fetchUserSubmittedData();

populateDropdown(martyrData, 'martyr-dropdown-menu');
populateDropdown(ipData, 'IP-dropdown-menu');
populateDropdown(userData, 'user-submitted-dropdown-menu');
```

File: portfolio/src/main/webapp/js/custom-script.js [hunk 19]
```diff
@@ -85,15 +85,6 @@ function codeAddress(address) {
...
   });
-  marker.addListener('click', toggleBounce);
-}
-
-function toggleBounce() {
-  if (marker.getAnimation() !== null) {
-    marker.setAnimation(null);
-  } else {
-    marker.setAnimation(google.maps.Animation.BOUNCE);
-  }
```
Comment (kennzoid) [inline portfolio/src/main/webapp/js/custom-script.js:96]: Why was this deleted?

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java (hunk 16 above)
Comment (kennzoid) [reply portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: +1

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java (hunk 15 above)
Comment (kennzoid) [reply portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: You'd have to submit the form using ajax instead to prevent a redirect. That could be a whole can of worms so I'm okay with a TODO here instead.

If you are curious, the answer here is an example of how to do that (ignore the php part):
https://stackoverflow.com/questions/5004233/jquery-ajax-post-example-with-php

**Review by kennzoid — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/index.html (hunk 17 above)
Comment (briafassler9) [reply portfolio/src/main/webapp/index.html]: +1. These changes were made in my previous PR requests.

File: portfolio/src/main/java/com/google/sps/servlets/MapServlet.java (hunk 2 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/MapServlet.java]: Changed

File: portfolio/src/main/java/com/google/sps/servlets/MapServlet.java (hunk 3 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/MapServlet.java]: Deleted.

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java (hunk 16 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: Edited to 'submitted-locations"

File: portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java (hunk 15 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserSubmittedCommentsServlet.java]: I added a TODO here and will come back and fix this shortly.

File: portfolio/src/main/webapp/js/custom-script.js (hunk 19 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/js/custom-script.js:96]: Since I instantiate the map with the drop animation within the codeAddress() function, a helper method isn't exactly necessary. The features still work as I intended without any of those lines, and it prevents lint checks from throwing "unused vars" errors.

File: portfolio/src/main/webapp/index.html (hunk 17 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/index.html]: It is diffing against the wrong base branch. I have the base of this PR on 'Comments' because a lot of the work done here is based on the things I coded within that PR.

File: portfolio/src/main/webapp/js/custom-script.js (hunk 18 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/js/custom-script.js]: Fixed

File: portfolio/src/main/webapp/js/custom-script.js (hunk 12 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/js/custom-script.js]: I have changed the name of the endpoint.

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+  public UserComment(
+      String name, String phone, String location, String description, long timeStamp, long id) {
+    this.name = name;
+    this.phone = phone;
+    this.location = location;
+    this.description = description;
+    this.timeStamp = timeStamp;
+    this.id = id;
+  }
+}
```
Comment (Jaxobc) [inline portfolio/src/main/java/com/google/sps/servlets/UserComment.java]: This seems to be marked as an issue.

File: portfolio/src/main/webapp/about.html [hunk 21]
```diff
@@ -157,6 +157,14 @@ <h3 class="section-subheading text-muted">We'd love to hear from you!</h3>
...
       </div>
     </section>
+    <div id="googleForm">
+      <iframe
+        src="https://docs.google.com/forms/d/e/1FAIpQLSfN4mh5SLDzB2RXp2zMNTi_GSD-ywjBPOG4Scws5TKR7FlWcA/viewform?embedded=true"
+        title="Google Feedback Form"
+      >
+        Loading…
+      </iframe>
+    </div>
```
Comment (Jaxobc) [inline portfolio/src/main/webapp/about.html]: I think this shouldn't be here.

**Review by Jaxobc — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/about.html (hunk 21 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/about.html]: The unedited embedded link from the google form: 
<iframe src="https://docs.google.com/forms/d/e/1FAIpQLSfN4mh5SLDzB2RXp2zMNTi_GSD-ywjBPOG4Scws5TKR7FlWcA/viewform?embedded=true" width="640" height="1660" frameborder="0" marginheight="0" marginwidth="0">Loading…</iframe>

I believe the loading is there in place of the form before it is fully loaded on the page to prevent a blank space from showing up.

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java (hunk 20 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserComment.java]: Could you expand please?

Comment (Jaxobc) [reply portfolio/src/main/java/com/google/sps/servlets/UserComment.java]: Github indicated that there should be another line at the end of this file.

File: portfolio/src/main/webapp/about.html (hunk 21 above)
Comment (Jaxobc) [reply portfolio/src/main/webapp/about.html]: I thought the Google Form was added in a separate PR.

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java (hunk 20 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserComment.java]: Ah i see. fixing now.

File: portfolio/src/main/webapp/about.html (hunk 21 above)
Comment (kofimeighan) [reply portfolio/src/main/webapp/about.html]: It was, but the base branch was to this branch

Comment (Jaxobc) [reply portfolio/src/main/webapp/about.html]: You can rebase if need be. 
https://gist.github.com/loilo/930f141d9acf89e9e734ffa042acd750
After following those instructions you can force push the PR with the relevant commits.

**Review by wangtong-g — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+public final class UserComment {
```
Comment (kennzoid) [inline portfolio/src/main/java/com/google/sps/servlets/UserComment.java:17]: Should this file and related classes be called UserSubmittedLocation now? Feel free to just TODO that as well, I know this PR has been hanging around a while.

**Review by kennzoid — APPROVED** [review]

**Review by briafassler9 — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/UserComment.java (hunk 22 above)
Comment (kofimeighan) [reply portfolio/src/main/java/com/google/sps/servlets/UserComment.java:17]: I think UserComment is most appropriate for this class because it stores more than just location. It stores name, phone #, location, and description so calling it UserLocation or something along those lines would be confusing.
